### ISSUE TITLE
Design for synchronizing object stores across clusters

### ISSUE THREAD

**Issue by travisn** [rook/rook#1584]
The object store has the capability to sync to another object store in the same cluster or another cluster in a different geographical region. See the [Ceph docs](http://docs.ceph.com/docs/master/radosgw/multisite/). Rook needs to enhance the object store CRDs to enable this scenario to seamlessly replicate between object stores.

Rook currently creates a realm, zone group, zone, and new pools for each new object store. Each of those has the same name as the object store, and the pools include the name of the object store. In this way, Rook can easily support multiple, independent object stores.

As a proof of concept, we can use the Rook toolbox to execute the commands needed to set up the multi-site replication. From there, we can design what it should look like with CRDs.

**Comment by caitlinbestler** [comment]
This is an example of a service where we would need for multiple storage backends to each advertise their capabilities. With most storage backends there is probably an inter-site replication method available which will be more efficient than rook applying generic replication code. But you still want there to be a generic fallback algorithm that can be applied when doing sync between any two arbitrary storage clusters.

**Comment by hunter** [comment]
Related to #1588 which might be worth considering for designing more general replication configuration/APIs

**Comment by brian-maloney** [comment]
@travisn I just recently manually configured Rook for multisite successfully using the ceph docs (and lots of radosgw-admin commands from the toolbox).  The RGWs needed NodePort and Ingress set up so they could communicate with each other in my config since they were in different clusters.

The trick besides manual config per the ceph docs was to exec into the rgw instance and edit /etc/ceph/ceph.conf with updated `rgw_zone` and `rgw_zonegroup` values.  Then within the container I ran `kill 1` and let k8s restart the container - /etc/ceph/ceph.conf was preserved (to my surprise) and the rgws were able to communicate with each other.

Obviously I would love to have the operator do all this for me, but in the meantime just allowing the ability to supply the zone and zonegroup names to the rgw might be sufficient for most use cases.

**Comment by galexrt** [comment]
@brian-maloney The manual changes to `ceph.conf` should be overriden by the init container(s) right now at Pod "restart" (e.g., when Kubelet is recreating Pod, normal Pod deletion).
Have you tried to use the Rook Ceph Custom Config override https://rook.io/docs/rook/v0.9/advanced-configuration.html#custom-cephconf-settings ?

**Comment by brian-maloney** [comment]
@galexrt Yep, that did the trick!  Thanks for the hint, I'm new to Rook and haven't read all the docs yet.  Now I can scale my RGW deployment down to 0 and back up and the RGWs come back up and sync resumes automatically.

**Comment by Negashev** [comment]
@travisn @galexrt 
Will there be any documentation? or promotion?

**Comment by travisn** [comment]
@Negashev This is in the [roadmap](https://github.com/rook/rook/blob/master/ROADMAP.md), although still needs an owner to drive it.

**Comment by BarthV** [comment]
:+1:

**Comment by mmgaggle** [comment]
Insofar as config override. How are we supporting different overrides for different clusters?

**Comment by leseb** [comment]
Based on recent discussions, it seems @alimaredia is going to give this one a try.

**Comment by michael-db** [comment]
@travisn "We've had some users attempt to follow the ceph docs without any success yet." You can add me to the list. I've created cephobjectstores on each side, then done a lot of experimenting in the toolbox pods, including adding the external NodePort URLs to the endpoints of the zone and zongroup and putting the Rook-created zone on the secondary side into the zonegroup pulled from the master side. in the year since you made that comment, have instructions been written down anywhere to get this working with Rook as it stands?
Or @brian-maloney, by any chance would you have a more detailed description of what you did?

**Comment by travisn** [comment]
The [design doc](https://github.com/rook/rook/blob/master/design/ceph/object/ceph-multisite-overview.md) for this feature has been completed with #4520. Leaving this issue opened, but renaming for implementation in v1.4.

**Comment by KlavsKlavsen** [comment]
@brian-maloney I have been trying to read on how to BEST setup rook w/ceph in multi-site setup - and examples are hard to find :(
Are you running a k8s cluster that spans the 2 sites? and then 2x Rook instances.. or is Rook also spanning the two sites and then how are you ensuring pods with "the active PVC" on site1 - stays on a node in site1 ? (and that failover to site2 works and pods rolled up on a site2 node, then uses the active PVC on site2 (the mirror) ?

Or are you running 2xRook's in same cluster (and then how does failover work - f.ex. for a mysql or postgresql with a master in site1 and slave in site2) ?
Or are you running 2x k8s clusters as well (one in each site).. ?
So many questions.. I am hoping you can share what works for you :)

**Comment by brian-maloney** [comment]
@KlavsKlavsen You would not want to build a single k8s cluster (or any kind of cluster that uses a RAFT-style consensus algorithm for cluster quorum, for example Zookeeper also suffers from this limitation) that spans regions, because those algorithms require short round-trip time.  In the environment I was working in at the time, we had 3 global regions (NA, EU, APAC) and had a k8s cluster in each of those regions.  Each cluster had its own Rook install, and data was replicated to all of those clusters.   We were not using Rook for PersistentVolumes/PVCs, just as an object store, but I imagine it would have worked for that use case too.
